### PR TITLE
[Dep] Upgrade flashinfer to 0.6.14

### PR DESCRIPTION
**Pull request by Fridge003**
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #29049902482](https://github.com/sgl-project/sglang/actions/runs/29049902482)
Latest PR Test (Extra): :white_check_mark: [Run #29049902443](https://github.com/sgl-project/sglang/actions/runs/29049902443)

### CODE REVIEW

**Comment by Fridge003** [comment]
Blocked by releasing of flashinfer-cubin wheel

**Comment by mmangkad** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_tp_mtp.py test/registered/cuda_graph/piecewise/test_pcg_glm52_fp4.py

**Comment by leejnau** [comment]
Maybe we should also update these for the new cubin install path and version-bump logic?

- docs_new/docs/get-started/install.mdx:220
- docs_new/docs/get-started/quickstart.mdx:293-297
- scripts/release/bump_flashinfer_version.py

**Comment by b8zhong** [comment]
@leejnau We don't need to change that, since for most cases (when using pip), it's OK for the cubin to be downloaded during the server launch anyway

File: docker/Dockerfile [hunk 1]
```diff
@@ -19,7 +19,7 @@ ARG PIP_DEFAULT_INDEX
 ARG UBUNTU_MIRROR
 ARG GITHUB_ARTIFACTORY=github.com
 ARG INSTALL_FLASHINFER_JIT_CACHE=0
-ARG FLASHINFER_VERSION=0.6.12
+ARG FLASHINFER_VERSION=0.6.14
```
Comment (Fridge003) [inline docker/Dockerfile:22]: Maybe update documents for flashifner-jitcache/cubin installation

File: python/pyproject.toml [hunk 2]
```diff
@@ -31,7 +31,10 @@ dependencies = [
   "einops",
   "fastapi",
   "flash-attn-4==4.0.0b15",
-  "flashinfer_cubin==0.6.14",
+  # TEMP workaround: flashinfer-cubin 0.6.14 is not on PyPI yet
+  # (flashinfer-ai/flashinfer#3808), so install it from the GitHub release.
+  # Switch back to flashinfer_cubin==0.6.14 once the wheel lands on PyPI.
+  "flashinfer_cubin @ https://github.com/flashinfer-ai/flashinfer/releases/download/v0.6.14/flashinfer_cubin-0.6.14-py3-none-any.whl",
```
Comment (Fridge003) [inline python/pyproject.toml]: Why we are not doing this

Comment (Fridge003) [reply python/pyproject.toml]: Maybe for future

File: docker/Dockerfile [hunk 3]
```diff
@@ -342,6 +342,10 @@ RUN --mount=type=cache,target=/root/.cache/pip \
         *) echo "Unsupported CUDA version: $CUDA_VERSION" && exit 1 ;; \
     esac \
     && mkdir -p /flashinfer_jit_output \
+    # flashinfer-cubin is CUDA-version-agnostic, unlike jit-cache, so its index-url has no cu${CUINDEX} suffix
+    && python3 -m pip install flashinfer-cubin==${FLASHINFER_VERSION} --index-url https://flashinfer.ai/whl \
+    && cp -r /usr/local/lib/python3.12/dist-packages/flashinfer_cubin /flashinfer_jit_output/ \
```
Comment (Fridge003) [inline docker/Dockerfile:347]: Why are we doing this? This will increase image size @mmangkad

File: test/registered/moe/test_cutedsl_moe.py [hunk 4]
```diff
@@ -627,22 +620,36 @@ def test_v2_cuda_graph_parity(self):
 
                 with torch.no_grad():
                     out_no_graph = _run_wrapper(wrapper_no_graph, tensors)
-                    out_graph = _run_wrapper(wrapper_graph, tensors)
-                    out_graph2 = _run_wrapper(wrapper_graph, tensors)
+
```
Comment (Fridge003) [inline test/registered/moe/test_cutedsl_moe.py:623]: Why modifying this test @mmangkad

File: docker/Dockerfile (hunk 3 above)
Comment (mmangkad) [reply docker/Dockerfile:347]: @Fridge003 this is not my change

cc @b8zhong

File: test/registered/moe/test_cutedsl_moe.py (hunk 4 above)
Comment (mmangkad) [reply test/registered/moe/test_cutedsl_moe.py:623]: @Fridge003 this is not my change

cc @b8zhong

File: python/pyproject.toml (hunk 2 above)
Comment (mmangkad) [reply python/pyproject.toml]: @Fridge003 @b8zhong removed this, but actually I prefer this so source/wheel installs still pull `flashinfer_cubin` automatically

File: docker/Dockerfile (hunk 3 above)
Comment (b8zhong) [reply docker/Dockerfile:347]: Cubin was already included by pip. We just do it to be the same as before

File: test/registered/moe/test_cutedsl_moe.py (hunk 4 above)
Comment (b8zhong) [reply test/registered/moe/test_cutedsl_moe.py:623]: Flashinfer refactored the wrappper (for cuda graph). https://github.com/flashinfer-ai/flashinfer/pull/3404 (now no way to pass preallocaetd buffer if not in cuda graph)
